### PR TITLE
feat: wait group

### PR DESCRIPTION
**Pull request by alvynben**
This PR introduces comprehensive support for the WaitGroup synchronization primitive in the go-slang interpreter, enhancing the concurrency capabilities of the language. Key components of the update include:

## Type System Enhancements:

1. New types (NewType and MakeType) have been added to accommodate the WaitGroup and differentiate between creation methods (new vs. make).
2. Modifications to the type system are reflected across type literals and command structures, expanding the interpreter's understanding of these new constructs.

## Interpreter Functionality:

1. The CallExpression handling in the interpreter now recognizes and processes methods called on WaitGroup instances (add, done, and wait), using a new structured approach to manage method-specific actions dynamically.
2. New operations for WaitGroup (WaitGroupAddOp, WaitGroupDoneOp, WaitGroupWaitOp) are implemented to handle the respective functionalities of the WaitGroup methods.

## Heap Management:

1. The heap management system has been updated to allocate and manage memory for WaitGroup instances correctly.
2. New tags (PointerTag.WaitGroup) have been added to manage WaitGroup pointers specifically.

## Video Description:

https://github.com/shenyih0ng/go-slang/assets/71116618/92c3c171-c5d9-463d-8b9f-955cf21d3c1b

### CODE REVIEW

File: src/go-slang/lib/waitgroup.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+  protected getCountAddr(): number { return WORD_SIZE }
+
+  protected getCount(): number { return this.memory.getFloat64(this.getCountAddr()) }
+
+  protected setCount(value: number): void { this.memory.setFloat64(this.getCountAddr(), value) }
+
+  public add(n: number): void { this.setCount(this.getCount() + n) }
+
+  public done(): void { this.setCount(this.getCount() - 1) }
```
Comment (alvynben) [inline src/go-slang/lib/waitgroup.ts]: NOTE: I might need to prevent count from going below 0.

File: src/go-slang/lib/heap/index.ts [hunk 2]
```diff
@@ -359,6 +372,15 @@ export class Heap {
     this.memory.setUint8(ptr_heap_addr + 3, 0) // initialize buffer size to 0
     return ptr_heap_addr
   }
+  /* Memory Layout of a WaitGroup: 
+   * [0-7:tag][0-7:count] (2 words) 
+   */
+  public allocateWaitGroup(): HeapAddress {
+    const ptr_heap_addr = this.allocateTaggedPtr(PointerTag.WaitGroup, 2)
```
Comment (shenyih0ng) [inline src/go-slang/lib/heap/index.ts]:
```suggestion
    const ptr_heap_addr = this.allocateTaggedPtr(PointerTag.WaitGroup, 1)
```

The `allocateTaggedPtr` takes in the size of the underlying data structure, so it implicitly +1 to account for the size of the tag heap object

File: src/go-slang/lib/heap/index.ts [hunk 3]
```diff
@@ -359,6 +372,15 @@ export class Heap {
     this.memory.setUint8(ptr_heap_addr + 3, 0) // initialize buffer size to 0
     return ptr_heap_addr
   }
+  /* Memory Layout of a WaitGroup: 
+   * [0-7:tag][0-7:count] (2 words)
```
Comment (shenyih0ng) [inline src/go-slang/lib/heap/index.ts]: Should we just use 1 word instead? We could keep the count in bytes 1:4 (2^32) in the first word instead using a additional word

File: src/go-slang/lib/waitgroup.ts (hunk 1 above)
Comment (shenyih0ng) [reply src/go-slang/lib/waitgroup.ts]: According to the specs, it panics when counter goes below 0. We could have the ECE raises a runtime error

Ref: https://pkg.go.dev/sync#WaitGroup.Add

Comment (alvynben) [reply src/go-slang/lib/waitgroup.ts]: Good idea.

File: src/go-slang/lib/heap/index.ts (hunk 2 above)
Comment (alvynben) [reply src/go-slang/lib/heap/index.ts]: Reasonable.

File: src/go-slang/lib/heap/index.ts (hunk 3 above)
Comment (alvynben) [reply src/go-slang/lib/heap/index.ts]: Will do.

File: src/go-slang/goroutine.ts [hunk 4]
```diff
@@ -293,14 +298,43 @@ const Interpreter: {
...
+    }
+
+    if (className instanceof WaitGroup) {
+      const methodActions = {
+        Add: () => ({ type: CommandType.WaitGroupAddOp, count: (args[0] as Literal).value }),
+        Done: () => ({ type: CommandType.WaitGroupDoneOp }),
+        Wait: () => ({ type: CommandType.WaitGroupWaitOp })
+      }
+
+      const action = methodActions[callee.method.name]()
```
Comment (shenyih0ng) [inline src/go-slang/goroutine.ts]: We will need to check if `callee.method.name` exists in the `methodActions.keys()` first. In the case where name does not exist, we will be calling a non-function

**Pull request merged**